cpp/src/arrow/error_or.h

```diff
@@ -0,0 +1,427 @@
+  void AssignStatus(U&& status) {
+    if (ok()) {
+      OverwriteValueWithStatus(std::forward<U>(status));
+    } else {
+      // Reuse the existing Status object. has_value_ is already false.
+      variant_.status_ = std::forward<U>(status);
+    }
+  }
+
+  // Under the assumption that |this| is currently holding a value, resets the
+  // |variant_| member to contain |status| and sets |has_value_| to indicate
+  // that |this| does not have a value. Destroys the existing |variant_| member.
+  template <class U>
+  void OverwriteValueWithStatus(U&& status) {
+#ifndef NDEBUG
+    if (!ok()) {
+      std::abort();
+    }
+#endif
+    variant_.value_.~T();
+    new (&variant_) variant(std::forward<U>(status));
+    has_value_ = false;
+  }
+
+  // Resets the |variant_| member to contain the |value| and sets |has_value_|
+  // to indicate that the ErrorOr object has a value. Destroys the existing
+  // |variant_| member.
+  template <class U>
+  void AssignValue(U&& value) {
+    if (ok()) {
+      // We cannot assume that T is move-assignable.
+      variant_.value_.~T();
+    } else {
+      variant_.status_.~Status();
+    }
+    new (&variant_) variant(std::forward<U>(value));
+    has_value_ = true;
+  }
+
+  // Use custom variant instead of c++ standard, to avoid header polution.
+  union variant {
+    // A non-OK status.
+    Status status_;
+
+    // An element of type T.
+    T value_;
+
+    variant() {}
+
+    variant(const Status& status) : status_(status) {}
+
+    variant(Status&& status) : status_(std::move(status)) {}
+
+    template <typename U, typename E = typename std::enable_if<
+                              std::is_constructible<T, U>::value &&
+                              std::is_convertible<U, T>::value>::type>
+    variant(U&& value) : value_(std::forward<U>(value)) {}
+
+    // This destructor must be explicitly defined because it is deleted due to
+    // the variant type having non-static data members with non-trivial
+    // destructors.
+    ~variant() {}
+  };
+
+  // One of: a non-OK status or an element of type T.
+  variant variant_;
+
+  // Indicates the active member of the variant_ member.
+  //
+  // A value of true indicates that value_ is the active member of variant_.
+  //
+  // A value of false indicates that status_ is the active member of variant_.
+  bool has_value_;
+};
+
+#define ASSIGN_OR_RAISE_IMPL(status_name, lhs, rexpr) \
+  auto status_name = (rexpr);                         \
+  RETURN_NOT_OK(status_name.status());                \
+  lhs = std::move(status_name).ValueOrDie();
+
+#define ASSIGN_OR_RAISE_NAME(x, y) ARROW_CONCAT(x, y)
+
+// Executes an expression that returns a ErrorOr, extracting its value
+// into the variable defined by lhs (or returning on error).
+//
+// Example: Assigning to an existing value
+//   ValueType value;
+//   ASSIGN_OR_RAISE(value, MaybeGetValue(arg));
+//
+// WARNING: ASSIGN_OR_RAISE expands into multiple statements; it cannot be used
```

Comment:
How can we make this more explicit? I suspect that a lot of code is written by using existing code as template (I'm guilty of this). This also requires the scope to return Status.

Reply:
FWIW, I don't think too many people will try to do this.  But I'm not sure I understand your comment fully.

Reply:
You should enclose the macro if a `do...while`, like this:
```cpp
ASSIGN_OR_RAISE_IMPL(status_name, lhs, rexpr) \
  do { \
    auto status_name = (rexpr);                         \
    ARROW_RETURN_NOT_OK(status_name.status());          \
    lhs = std::move(status_name).ValueOrDie(); \
  } while(0)
```

Reply:
This was intentional.  See macro usage, the intent is to allow lhs to escape scope (see usage in writer)  Perhaps there is better preprocessor magic to allow this though?

Reply:
As long as lhs escapes scope, why not have the macro declare it (instead of requiring users to do so?):

```c++
ARROW_VALUE_OR_RAISE_IMPL(status_name, lhs, rexpr) \
  decltype((rexpr).ValueOrDie()) lhs;
  do { \
    auto status_name = (rexpr);                         \
    ARROW_RETURN_NOT_OK(status_name.status());          \
    lhs = std::move(status_name).ValueOrDie(); \
  } while(0)
```

Reply:
Also if we add `Status::operator bool` we can do this in a way that doesn't require status_name:

```c++
ARROW_VALUE_OR_RAISE(lhs, rexpr)               \
  decltype((rexpr).ValueOrDie()) lhs;          \
  if (Status s = [&](decltype((rexpr)) r) {    \
    ARROW_RETURN_NOT_OK(r.status());           \
    lhs = std::move(r).ValueOrDie();           \
    return Status::OK();                       \
  }((rexpr)))                                  \
  return s
```

Reply:
@bkietz that is very impressive :)

See above about declaring bool, not quite working.

I think allowing the user to have control of whether the type is declared or they decide to use auto is important as it allows them decide which adds clarity. (https://google.github.io/styleguide/cppguide.html#auto)